### PR TITLE
KAFKA-16963: System test for adding and removing voters

### PR DESCRIPTION
**Pull request by ahuang98**
System test for quorum reconfiguration, combined and isolated mode

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tests/docker/ducker-ak
Comment (ahuang98) [inline tests/docker/ducker-ak]: won't merge this, fixes ducker up issue for me

File: tests/kafkatest/services/kafka/kafka.py
Comment (ahuang98) [inline tests/kafkatest/services/kafka/kafka.py]: will remove this comment before merging

File: tests/kafkatest/services/kafka/kafka.py
Comment (ahuang98) [inline tests/kafkatest/services/kafka/kafka.py:731]: this can be removed as start_node already does this

**Comment by ahuang98** [comment]
1 test failure which looks to be unrelated - `testStateGlobalThreadClose() – org.apache.kafka.streams.KafkaStreamsTest`

File: tests/kafkatest/services/kafka/kafka.py
Comment (cmccabe) [inline tests/kafkatest/services/kafka/kafka.py]: can we use a name like `dynamicRaftQuorum` here rather than the KIP number?

File: tests/kafkatest/services/kafka/kafka.py
Comment (cmccabe) [inline tests/kafkatest/services/kafka/kafka.py:638]: can we just put the controllers we want to skip into `nodes_to_skip`?

File: tests/kafkatest/services/kafka/kafka.py
Comment (cmccabe) [inline tests/kafkatest/services/kafka/kafka.py:908]: why are we not specifying the `--initial-voters` here? seems wrong (for clusters with more than one controller)

File: tests/kafkatest/services/kafka/templates/kafka.properties
Comment (cmccabe) [inline tests/kafkatest/services/kafka/templates/kafka.properties:28]: do we need an "if this is set, define it, otherwise not" here?

Comment (ahuang98) [reply tests/kafkatest/services/kafka/templates/kafka.properties:28]: I don't believe it's necessary, but I'll do so to make it clear that we may have either or

File: tests/kafkatest/services/kafka/kafka.py
Comment (ahuang98) [reply tests/kafkatest/services/kafka/kafka.py:908]: `--initial-controllers`? I think either way is correct based on [Jose's docs](https://github.com/apache/kafka/pull/17076/files#diff-e4e8d893dc2a4e999c96713dd5b5857203e0756860df0e70fb0cb041aa4d347bR3806). I would say not using it represents the point of our test a bit better - we initially wanted one controller but then later decide to add another one

File: tests/kafkatest/services/kafka/kafka.py
Comment (ahuang98) [reply tests/kafkatest/services/kafka/kafka.py:638]: hm, feel free to correct me if I'm wrong - it doesn't seem like it will work because we specifically rely on `self.kafka.start()` for our test (which will also create and start the isolated controller quorum as part of the startup process) and `self.kafka.start()` already is using `nodes_to_skip` to understand which broker-only nodes it needs to skip startup for in this scenario

Comment (ahuang98) [reply tests/kafkatest/services/kafka/kafka.py:638]: it seems a bit messy to change the meaning of `nodes_to_skip` to mean "skip isolated controller nodes instead of broker nodes if in isolated mode", so I decided to just add another parameter to make it more clear

File: tests/kafkatest/services/kafka/kafka.py
Comment (jsancio) [inline tests/kafkatest/services/kafka/kafka.py:208]: Let's document this new parameter in the doc string below.

File: tests/kafkatest/services/kafka/kafka.py
Comment (jsancio) [inline tests/kafkatest/services/kafka/kafka.py:912]: Shouldn't this be set in the same if statement block as where you add the `--standalone` flag to the command. In other words you want to guarantee that only one controller is bootstrapped with `--standalone`. On that note, how about naming this field `first_controller_bootstrapped_standalone` or something like that.

File: tests/kafkatest/services/kafka/kafka.py
Comment (jsancio) [inline tests/kafkatest/services/kafka/kafka.py:1804]: Is the dictionary needed since there is only one parameter to "format". How about:
```python
        cmd += "%s describe --status" % self.kafka_metadata_quorum_cmd(node)
```

Having said that, my understanding is that Python has moved away from using `%` for string formatting in favor of `string.format` and f-string.

File: tests/kafkatest/services/kafka/kafka.py
Comment (jsancio) [inline tests/kafkatest/services/kafka/kafka.py:1834]: Can we settle with one string formatter? In this method you are using both f-string and `%` formatting. This makes it very difficult to read in my opinion.

File: tests/kafkatest/services/kafka/kafka.py
Comment (jsancio) [inline tests/kafkatest/services/kafka/kafka.py:1848]: Btw, it looks like logger supports `string.format`.
```python
        self.logger.info("Running remove controller command...\n%s", cmd)
```

File: tests/kafkatest/services/kafka/kafka.py
Comment (jsancio) [inline tests/kafkatest/services/kafka/kafka.py:464]: Why is this correct? Can we document the idea behind this computation? At first glance it is not clear to me why this is correct.

File: tests/kafkatest/services/kafka/kafka.py
Comment (jsancio) [inline tests/kafkatest/services/kafka/kafka.py:877]: typo:
```python
            # define controller.quorum.bootstrap.servers or controller.quorum.voters text
```

File: tests/kafkatest/services/kafka/kafka.py
Comment (jsancio) [reply tests/kafkatest/services/kafka/kafka.py:908]: I think that the system tests that @ahuang98 is adding always start with one controller and expand or shrink the cluster as required by the test. When we convert the existing tests to use kraft.version 1 we will probably need to extend this to support `--initial-controllers`.

I am okay keeping this PR as is and added support for this case in a future PR.

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (jsancio) [inline tests/kafkatest/tests/core/quorum_reconfiguration_test.py:88]: Is there are specific reason why you are using 4 nodes? The replication factor is 3 so this can just use 3 nodes/brokers, right?

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (jsancio) [inline tests/kafkatest/tests/core/quorum_reconfiguration_test.py:84]: We need 6 nodes because there are 4 broker, 2 controller combined with two of the brokers and 2 consumer/producers. Is this accurate?

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (jsancio) [inline tests/kafkatest/tests/core/quorum_reconfiguration_test.py:113]: Why `2:`? 0 is a broker and standalone controller. 1 is a broker and observer controller. 2 is a broker. 3 is a broker. Is this accurate?

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (jsancio) [inline tests/kafkatest/tests/core/quorum_reconfiguration_test.py:98]: You need to skip `inactive_controller` because the node are started concurrently and you want to guarantee that `self.kafka.nodes[0]` is the `--standalone` controller?

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (jsancio) [inline tests/kafkatest/tests/core/quorum_reconfiguration_test.py:86]: Let's write a high-level comment documenting what we are testing. I always found system test really dense and hard to read. It would be good to have a mini-guide when reading system tests.

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (jsancio) [inline tests/kafkatest/tests/core/quorum_reconfiguration_test.py:61]: How do we make sure that the tests don't have a race? Are you concerned that the broker/controller was started but it has yet to send a FETCH request to the active controller (kraft leader).

This comment applies to all of the describe quorum checks in this method.

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (jsancio) [inline tests/kafkatest/tests/core/quorum_reconfiguration_test.py:81]: `active_controller_id` should show in `CurrentObservers`, no?

File: tests/kafkatest/services/kafka/kafka.py
Comment (cmccabe) [reply tests/kafkatest/services/kafka/kafka.py:908]: I think this is something we should add since I feel that `--initial-voters` is actually going to be more common. But let's not block the PR on this.

File: tests/kafkatest/services/kafka/kafka.py
Comment (cmccabe) [reply tests/kafkatest/services/kafka/kafka.py:638]: ok.

**Review by cmccabe — APPROVED** [review]
LGTM

**Review by jsancio — COMMENTED** [review]
Thanks for the tests @ahuang98 . I am concerned that these system test may be very indeterministic (flaky). Left discuss offline on how we can fix this.

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (ahuang98) [reply tests/kafkatest/tests/core/quorum_reconfiguration_test.py:81]: Hm, that would make sense to me - from additional logging I can see node 1 (active_controller_id) is added to observer states at some point, but it seems the describe command never catches this state since node 1 resigns as active controller quickly and is never added back to observer states

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (ahuang98) [reply tests/kafkatest/tests/core/quorum_reconfiguration_test.py:98]: not exactly, I wanted the test to be more intentional in checking each step of adding a new controller is correct (e.g. describe quorum checks between each step). "skipping" startup for one of the nodes seemed an easy way to achieve this with the current framework

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (ahuang98) [reply tests/kafkatest/tests/core/quorum_reconfiguration_test.py:113]: this is intentional, but the parameter name "broker_ids" is misleading, perhaps "broker_only_nodes" is slightly better

File: tests/kafkatest/services/kafka/kafka.py
Comment (ahuang98) [reply tests/kafkatest/services/kafka/kafka.py:464]: I'll add a comment, this is how `KafkaService.__init__` defines node ids for controller-only nodes, I just moved it into a helper so I could reference it in my test as well.

File: tests/kafkatest/services/kafka/kafka.py
Comment (ahuang98) [reply tests/kafkatest/services/kafka/kafka.py:912]: makes sense, I had put it there to be more generic (first controller has been started, doesn't matter if it was started with `--standalone` or not), but I don't think it's needed. 

I was trying to safeguard against a scenario like the following, but someone would really need to be doing something weird with their test to achieve this.
T=0 `kafka-storage.sh format ... --cluster-id AAA`
T=1 `kafka-storage.sh format ... --standalone --cluster-id AAA`

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (ahuang98) [reply tests/kafkatest/tests/core/quorum_reconfiguration_test.py:88]: 4 nodes is the max number of nodes we will get to - we start the test with 3 nodes and end the test with 3 nodes (which is the min needed for replication factor 3)

File: tests/kafkatest/tests/core/quorum_reconfiguration_test.py
Comment (ahuang98) [reply tests/kafkatest/tests/core/quorum_reconfiguration_test.py:84]: correct!

**Pull request merged**
